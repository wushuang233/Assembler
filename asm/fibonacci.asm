# 计算斐波那契数列的前7个数字 (1, 1, 2, 3, 5, 8, 13)
# 使用寄存器：
# x1: 当前数
# x2: 下一个数
# x3: 临时存储用于计算
# x4: 计数器
# x5: 存储目标数列长度 (7)

addi x1, x0, 1    # 初始化第一个斐波那契数 f(1) = 1
addi x2, x0, 1    # 初始化第二个斐波那契数 f(2) = 1
addi x4, x0, 2    # 计数器从2开始（已经计算了两个数）
addi x5, x0, 7    # 目标是计算7个斐波那契数

add x3, x1, x2    # x3 = x1 + x2 (计算下一个斐波那契数)
add x1, x0, x2    # x1 = x2 (更新当前数)
add x2, x0, x3    # x2 = x3 (更新下一个数)
addi x4, x4, 1    # 计数器+1
bne x4, x5, -16   # 如果计数器 != 7，返回到第一条add指令
halt              # 程序结束

# 当循环结束时，x2中存储的是第7个斐波那契数 (13) 