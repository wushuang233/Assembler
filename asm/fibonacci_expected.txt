# 计算斐波那契数列的前7个数字 (1, 1, 2, 3, 5, 8, 13)
# addi x1, x0, 1 (R1 = 1)
0b00000000000_00001_00000_00001_000010
# addi x2, x0, 1 (R2 = 1)
0b00000000000_00001_00000_00010_000010
# addi x4, x0, 2 (R4 = 2)
0b00000000000_00010_00000_00100_000010
# addi x5, x0, 7 (R5 = 7)
0b00000000000_00111_00000_00101_000010
# add x3, x1, x2 (R3 = R1 + R2)
0b00000000000_00010_00001_00011_000001
# add x1, x0, x2 (R1 = R2)
0b00000000000_00010_00000_00001_000001
# add x2, x0, x3 (R2 = R3)
0b00000000000_00011_00000_00010_000001
# addi x4, x4, 1 (R4 = R4 + 1)
0b00000000000_00001_00100_00100_000010
# bne x4, x5, -16 (如果R4 != R5，跳回16字节)
0b11111111111_00101_00100_10000_000011
# halt
0b00000000000_00000_00000_00000_000000 